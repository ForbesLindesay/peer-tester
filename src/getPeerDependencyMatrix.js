import getVersions from './getVersions';

async function getPeerDependencyMatrix(dependencies) {
  const versions = await Promise.all(
    Object.keys(dependencies).sort().map(async (name) => {
      return [name, await getVersions(name, dependencies[name])];
    }),
  );
  return toMatrix(versions);
}

function toMatrix(list) {
  if (list.length === 0) {
    return [{}];
  }
  const [name, versions] = list[0];
  const result = [];
  for (const child of toMatrix(list.slice(1))) {
    for (const version of versions) {
      result.push({...child, [name]: version});
    }
  }
  return result;
}

export default getPeerDependencyMatrix;
