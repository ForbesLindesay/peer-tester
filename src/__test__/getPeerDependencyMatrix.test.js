import getPeerDependencyMatrix from '../getPeerDependencyMatrix';

test('getPeerDependencyMatrix({"promise": "^6.0.0", "then-request": "^1.0.0"})', async () => {
  const matrix = await getPeerDependencyMatrix({"promise": "^6.0.0", "then-request": "^1.0.0"});
  expect(matrix).toMatchSnapshot();
});
