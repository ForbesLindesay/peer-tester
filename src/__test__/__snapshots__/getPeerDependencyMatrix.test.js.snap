// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getPeerDependencyMatrix({"promise": "^6.0.0", "then-request": "^1.0.0"}) 1`] = `
Array [
  Object {
    "promise": "6.0.0",
    "then-request": "1.0.0",
  },
  Object {
    "promise": "6.0.1",
    "then-request": "1.0.0",
  },
  Object {
    "promise": "6.1.0",
    "then-request": "1.0.0",
  },
  Object {
    "promise": "6.0.0",
    "then-request": "1.0.1",
  },
  Object {
    "promise": "6.0.1",
    "then-request": "1.0.1",
  },
  Object {
    "promise": "6.1.0",
    "then-request": "1.0.1",
  },
  Object {
    "promise": "6.0.0",
    "then-request": "1.0.2",
  },
  Object {
    "promise": "6.0.1",
    "then-request": "1.0.2",
  },
  Object {
    "promise": "6.1.0",
    "then-request": "1.0.2",
  },
  Object {
    "promise": "6.0.0",
    "then-request": "1.0.3",
  },
  Object {
    "promise": "6.0.1",
    "then-request": "1.0.3",
  },
  Object {
    "promise": "6.1.0",
    "then-request": "1.0.3",
  },
  Object {
    "promise": "6.0.0",
    "then-request": "1.0.4",
  },
  Object {
    "promise": "6.0.1",
    "then-request": "1.0.4",
  },
  Object {
    "promise": "6.1.0",
    "then-request": "1.0.4",
  },
  Object {
    "promise": "6.0.0",
    "then-request": "1.0.5",
  },
  Object {
    "promise": "6.0.1",
    "then-request": "1.0.5",
  },
  Object {
    "promise": "6.1.0",
    "then-request": "1.0.5",
  },
  Object {
    "promise": "6.0.0",
    "then-request": "1.1.0",
  },
  Object {
    "promise": "6.0.1",
    "then-request": "1.1.0",
  },
  Object {
    "promise": "6.1.0",
    "then-request": "1.1.0",
  },
]
`;
